### ISSUE TITLE
Errors using method decorator with generic class

### ISSUE THREAD

**Issue by bryanforbes** [microsoft/pyright#6574]
**Describe the bug**
Given the following code:

```python
from __future__ import annotations

from collections.abc import Callable
from typing import Any, Concatenate, Generic, ParamSpec, TypeAlias
from typing_extensions import TypeVar

_T = TypeVar('_T', infer_variance=True)
_P = ParamSpec('_P')
_R = TypeVar('_R', infer_variance=True)

_F: TypeAlias = Callable[Concatenate['Context[Any]', _P], _R]


def ensure_db(func: _F[_P, _R], /) -> _F[_P, _R]:
    ...


class Context(Generic[_T]):
    @ensure_db
    def select_all(self, /, *args: object) -> list[Any]:
        ...
```

An error is generated for line 19:

```
blah.py
  blah.py:19:6 - error: Argument of type "(self: Self@Context[_T@Context], /, *args: object) -> list[Any]" cannot be assigned to parameter "func" of type "_F[_P@ensure_db, _R@ensure_db]" in function "ensure_db"
    Type "(self: Self@Context[_T@Context], /, *args: object) -> list[Any]" cannot be assigned to type "_F[_P@ensure_db, _R@ensure_db]"
      Parameter 1: type "Context[Any]" cannot be assigned to type "Self@Context[_T@Context]"
        Type "Context[Any]" cannot be assigned to type "Self@Context[_T@Context]" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
```

If I try to fix the above error by modifying the code to the following:

```python
from __future__ import annotations

from collections.abc import Callable
from typing import Any, Concatenate, Generic, ParamSpec, TypeAlias
from typing_extensions import TypeVar

_T = TypeVar('_T', infer_variance=True)
_C = TypeVar('_C', bound='Context[Any]', infer_variance=True)
_P = ParamSpec('_P')
_R = TypeVar('_R', infer_variance=True)

_F: TypeAlias = Callable[Concatenate[_C, _P], _R]


def ensure_db(func: _F[_C, _P, _R], /) -> _F[_C, _P, _R]:
    ...


class Context(Generic[_T]):
    @ensure_db
    def select_all(self, /, *args: object) -> list[Any]:
        ...
```

The following error is generated on line 19:

```
blah.py
  blah.py:20:6 - error: Expected 0 positional arguments (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool?
Command Line

Which version?
1.1.338

**Comment by erictraut** [comment]
The issue in your first code sample is not a bug. It's an intentional change made in 1.1.338 to plug a previous hole in type checking.

The workaround in your second code sample is the one I was going to suggest, and I ran into the same problem you did as I was composing my suggestion.

The error in the second code sample is a false positive, and I'm working on a fix. I've spent some time on it already, but I don't yet fully understand the cause. I have determined that it's a long-standing bug, not a recent regression.

I've also found some workarounds, which you may want to use until I have a fix in place:
1. Switch from a `TypeAlias` to a PEP-695 `type` statement to define `_F`.
2. In your definition of `ensure_db`, manually expand the first use of the type alias `_F` to `Callable[Concatenate[_X, _P], _R]`.

**Comment by erictraut** [comment]
I just figured out the problem. It's caused by a complex interaction between inferred variance, decorators, and generic type aliases. That explains why none of our current unit tests caught this problem previously.

Now that I fully understand the issue, I can offer another workaround. You can remove the `infer_variance=True` from `_T`. (The `infer_variance` that you're using for `R` is meaningless because `_R` is not used as a class type parameter in your code sample.)

**Comment by erictraut** [comment]
This will be fixed in the next release. I've added a variation of your code sample to our unit tests.

**Comment by bryanforbes** [comment]
Thank you, @erictraut!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.339, which I just published. It will also be included in a future release of pylance.
